ibis/expr/analysis.py

```diff
@@ -30,29 +19,39 @@
 
 
 def sub_for(expr, substitutions):
-    mapping = {repr(k.op()): v for k, v in substitutions}
-    substitutor = Substitutor()
-    return substitutor.substitute(expr, mapping)
+    """Substitute subexpressions in `expr` with expression to expression
+    mapping `substitutions`.
 
+    Parameters
+    ----------
+    expr : ibis.expr.types.Expr
+        An Ibis expression
+    substitutions : Mapping[ibis.expr.types.Expr, ibis.expr.types.Expr]
```

Comment:
Based on the dict comprehension below, `substitutions` seems like a list of 2-tuples.